### ISSUE TITLE
Weak reference support

### ISSUE THREAD

**Issue by nogayama** [netaddr/netaddr#161]
Since I am making many IPAddress objects, I want to cache them.
Sharing an IPAddress object with several strings that indicate the same ip address can reduce memory footprint. 

For example, I made the following helper method.

```python
from netaddr.ip import IPAddress
import weakref
str2ipobj = weakref.WeakValueDictionary()
def create_ip(ip_str):
    if ip_str in str2ipobj:
        return str2ipobj[ip_str]
    else:
        ip_obj = IPAddress(ip_str)
        str2ipobj[ip_str] = ip_obj
        return ip_obj
```

But this did not work with latest version of netaddr 0.7.19. The error message was 

```
cannot create weak reference to 'IPAddress' object
```

**Comment by jstasiak** [comment]
This is interesting, I'll have to read some documentation, I thought types support weak references out of the box.

**Comment by jstasiak** [comment]
This has been fixed in 0.8.0.
